File: clients/typescript/test/service-smoke.mjs
// Explicit actual-service canary. Creates and deletes only its own named sandbox.
// BOG_TEST_ORIGIN and BOG_TEST_AUTH_FILE must identify an authorized test service.
import {Client,installPrivate} from '../src/index.js';
import {mkdtemp,readFile,rm} from 'node:fs/promises';
import {tmpdir} from 'node:os';
import {join} from 'node:path';
import {randomUUID} from 'node:crypto';
import assert from 'node:assert/strict';
const base=process.env.BOG_TEST_ORIGIN,authFile=process.env.BOG_TEST_AUTH_FILE;
if(!base||!authFile)throw new Error('Set BOG_TEST_ORIGIN and private BOG_TEST_AUTH_FILE.');
const c=await Client.fromAuth(authFile,base,process.env.BOG_TEST_WORKSPACE_ID);
const temporary=await mkdtemp(join(tmpdir(),'bog-service-smoke-'));
try{
 const definition=JSON.parse(await readFile(new URL('../../../docs/examples/composable/todo-search.json',import.meta.url),'utf8'));
 const created=await c.create('client-smoke-'+randomUUID(),randomUUID(),definition,60,{sandbox:true,wait:true,appAccess:{scope:'read',label:'Client smoke'}});
 assert.equal(created.status,'ready');assert.ok(created.app_access.handoff_id);
 await installPrivate(base,authFile,created.app_access.handoff_id,join(temporary,'app.env'),'dotenv');
 const app=await Client.fromConfig(join(temporary,'app.env'));
 const before=(await c.wait(undefined,0)).cursor;
 await c.put('b',{title:'smoke alpha',notes:'findable',priority:9,completed:false});
 await c.put('c',{title:'smoke beta',notes:'findable',priority:3,completed:false});
 assert.equal((await c.wait(before,0)).changed,true);
 const batch=await app.batchGet(['b','missing','b']);assert.equal(batch.data.length,3);assert.equal(batch.data[1].value,null);assert.deepEqual(batch.data[0],batch.data[2]);
 const page=await app.list('docs',{after:'a',before:'d',limit:10});assert.ok(JSON.stringify(page).includes('smoke alpha'));
 const rank=await app.top('priority',10,0,['/title']);assert.equal(rank.data[0].value['/title'],'smoke alpha');
 const search=await app.search('text_search','smoke',3,['/title']);assert.ok(JSON.stringify(search).includes('/title'));
 await app.routes();await app.diagnostics();
 await assert.rejects(app.put('forbidden',{title:'no'}));
 console.log(JSON.stringify({status:'passed',checks:['sandbox-readiness','private-install-dotenv','batch-get-null-duplicates','range','rank','search-projection','change-wait','routes','diagnostics','read-scope']}));
}finally{
 try{if(c.bogId)await c.deleteSandbox(c.bogId);}finally{await rm(temporary,{recursive:true,force:true});}
}

File: clients/python/service_smoke.py
#!/usr/bin/env python3
"""Explicit actual-service smoke; creates and deletes only its own sandbox.

BOG_TEST_ORIGIN and BOG_TEST_AUTH_FILE select a private management authorization.
Only check names are printed; private app credentials stay in a temporary file.
"""
import contextlib
import json
import os
from pathlib import Path
import sys
import tempfile
import uuid
from bog_client import Client, BogError
from bog_app_access import main as install_private


def main():
    base=os.environ['BOG_TEST_ORIGIN'];auth=os.environ['BOG_TEST_AUTH_FILE']
    client=Client.from_auth(auth,base,os.environ.get('BOG_TEST_WORKSPACE_ID'))
    definition=json.loads((Path(__file__).resolve().parents[2]/'docs/examples/composable/todo-search.json').read_text())
    with tempfile.TemporaryDirectory(prefix='bog-service-smoke-') as directory:
        try:
            result=client.create('client-smoke-'+str(uuid.uuid4()),str(uuid.uuid4()),definition,sandbox=True,wait=True,app_access={'scope':'read','label':'Python client smoke'})
            assert result['status']=='ready'
            config=Path(directory)/'app.env'
            with contextlib.redirect_stdout(sys.stderr):
                install_private(['--origin',base,'--auth-file',auth,'--handoff',result['app_access']['handoff_id'],'--output',str(config),'--format','dotenv'])
            app=Client.from_config(config)
            cursor=client.wait(timeout=0)['cursor']
            client.put('b',{'title':'smoke alpha','notes':'findable','priority':9,'completed':False})
            client.put('c',{'title':'smoke beta','notes':'findable','priority':3,'completed':False})
            assert client.wait(cursor,0)['changed']
            batch=app.batch_get(['b','missing','b'])['data']
            assert len(batch)==3 and batch[1]['value'] is None and batch[0]==batch[2]
            page=app.list(after='a',before='d',limit=10)
            assert 'smoke alpha' in json.dumps(page)
            rank=app.top('priority',10,0,['/title'])
            assert rank['data'][0]['value']['/title']=='smoke alpha'
            search=app.search('text_search','smoke',3,['/title'])
            assert '/title' in json.dumps(search)
            app.routes();app.diagnostics()
            try:app.put('forbidden',{'title':'no'})
            except BogError as error:assert error.status==403
            else:raise AssertionError('read scope accepted write')
            print(json.dumps({'status':'passed','checks':['sandbox-readiness','private-install-dotenv','batch-get-null-duplicates','range','rank-projection','search-projection','change-wait','routes','diagnostics','read-scope']}))
        finally:
            if client.bog_id:client.delete_sandbox(client.bog_id)


if __name__=='__main__':
    try:main()
    except Exception:
        print('Service smoke failed; inspect the test service and retained request observations.',file=sys.stderr)
        sys.exit(1)
